// rc-range-slider/lib/shared/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var isArray = function isArray(array) {
  return Object.prototype.toString.call(array) === '[object Array]';
};

var isEqual = function isEqual(arrayA, arrayB) {
  for (var i = 0; i < arrayA.length; i++) {
    if (arrayA[i] !== arrayB[i]) {
      return false;
    }
  }
  return true;
};

var mousePosition = function mousePosition(e) {
  e = e || window.event;
  var xPos = e.pageX ? e.pageX : e.clientX + document.body.scrollLeft - document.body.clientLeft;
  var yPos = e.pageY ? e.pageY : e.clientY + document.body.scrollTop - document.body.clientTop;
  return {
    x: xPos,
    y: yPos
  };
};

var disableMouseDown = function disableMouseDown(e) {
  var event = e || window.event;
  event.preventDefault();
  event.stopPropagation();
};

var getStandardAbsolutePosition = function getStandardAbsolutePosition(position, minPosition, maxPosition) {
  if (position < minPosition) {
    position = minPosition;
  }
  if (position > maxPosition) {
    position = maxPosition;
  }
  return position;
};

exports.default = {
  isArray: isArray,
  isEqual: isEqual,
  mousePosition: mousePosition,
  disableMouseDown: disableMouseDown,
  validatePosition: getStandardAbsolutePosition
};
